crates/miden-objects/src/batch/input_output_note_tracker.rs

```diff
@@ -0,0 +1,381 @@
+
+use crate::{
+    batch::{BatchId, ProvenBatch},
+    block::BlockNumber,
+    errors::ProposedBatchError,
+    note::{NoteHeader, NoteId, NoteInclusionProof, Nullifier},
+    transaction::{ChainMmr, InputNoteCommitment, OutputNote, ProvenTransaction, TransactionId},
+    ProposedBlockError,
+};
+
+// INPUT OUTPUT NOTE TRACKER
+// ================================================================================================
+
+/// A helper struct to track input and output notes and erase those that are created and consumed
+/// within the same batch or block.
+///
+/// Its main purpose is to check for duplicates and allow for removal of output notes that are
+/// consumed in the same batch, so are not output notes of the batch.
+///
+/// The approach for this is that the output note set is initialized to the union of all output
+/// notes of the transactions in the batch.
+/// Then (outside of this struct) all input notes of transactions in the batch which are also output
+/// notes can be removed, as they are considered consumed within the batch and will not be visible
+/// as created or consumed notes for the batch.
+#[derive(Debug)]
+pub(crate) struct InputOutputNoteTracker<ContainerId> {
+    /// An index from Nullifier to the identifier that consumes it (either a [`TransactionId`] or
+    /// [`BatchId`](crate::batch::BatchId)).
+    input_notes: BTreeMap<Nullifier, (ContainerId, InputNoteCommitment)>,
+    /// An index from [`NoteId`]s to the transaction that creates the note and the note itself.
+    /// The transaction ID is tracked to produce better errors when a duplicate note is
+    /// encountered.
+    output_notes: BTreeMap<NoteId, (ContainerId, OutputNote)>,
+}
+
+impl InputOutputNoteTracker<TransactionId> {
+    /// Computes the input and output notes for a transaction batch from the provided iterator over
+    /// transactions. Implements batch-specific logic.
+    pub fn from_transactions<'a>(
+        txs: impl Iterator<Item = &'a ProvenTransaction> + Clone,
+        unauthenticated_note_proofs: &BTreeMap<NoteId, NoteInclusionProof>,
+        chain_mmr: &ChainMmr,
+    ) -> Result<(Vec<InputNoteCommitment>, Vec<OutputNote>), ProposedBatchError> {
+        let input_notes_iter = txs.clone().flat_map(|tx| {
+            tx.input_notes()
+                .iter()
+                .map(|input_note_commitment| (input_note_commitment.clone(), tx.id()))
+        });
+        let output_notes_iter = txs.flat_map(|tx| {
+            tx.output_notes().iter().map(|output_note| (output_note.clone(), tx.id()))
+        });
+
+        let mut tracker = Self::from_iter(
+            input_notes_iter,
+            output_notes_iter,
+            unauthenticated_note_proofs,
+            chain_mmr,
+        )
+        .map_err(ProposedBatchError::from)?;
+
+        let batch_input_notes = tracker.erase_notes().map_err(ProposedBatchError::from)?;
+
+        Ok((batch_input_notes, tracker.into_final_output_notes()))
```

Comment:
Similar comment as above. We can probably do the conversion from the `BTreeMap` into a `Vec` here.